jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java

```diff
@@ -754,8 +757,19 @@ public Action process() throws Exception
 
                     case NEED_HEADER:
                     {
-                        _header = _bufferPool.acquire(_config.getResponseHeaderSize(), HEADER_BUFFER_DIRECT);
-
+                        int size;
+                        if (_header == null)
+                        {
+                            size = Math.min(_config.getResponseHeaderSize(), _config.getOutputBufferSize());
+                        }
+                        else
+                        {
+                            if (_header.capacity() >= _config.getResponseHeaderSize())
+                                throw new BadMessageException(INTERNAL_SERVER_ERROR_500, "Response header too large");
+                            size = _config.getResponseHeaderSize();
+                            _bufferPool.release(_header);
+                        }
+                        _header = _bufferPool.acquire(size, HEADER_BUFFER_DIRECT);
```

Comment:
Shouldn't we be acquiring based on `_header.capacity()` or `_config.getOutputBufferSize()` instead here?

The issue that was filed is that we allocate a buffer based on `_config.getResponseHeaderSize()`, even if it's set to something massive (like 100MB)
Even though the actual buffer that is required/used is much smaller.

Reply:
On the first attempt we allocate on the minimum of responseHeaderSize and outputBufferSize, which almost always will be responseHeaderSize, so if we do a second attempt the capacity test will be >= responseHeaderSize and we throw bad message.

Only if the responseHeaderSize is larger than the outputBufferSize do we first allocate the outputBufferSize and if we overflow we do a second attempt, which passes the capacity test and reallocates the buffer at responseHeaderSize (unless of course the buffer pool already gave us buffer that was big enough for that).

Reply:
See my comment above. If we introduce `NEED_LARGER_BUFFER`, this `if/else` can be split in 2 `case` blocks and be clearer.